### PR TITLE
Adding push-to-gcs action and README

### PR DESCRIPTION
**Pull request by ricky-undeadcoders**

### CODE REVIEW

File: actions/push-to-gcs/action.yaml
Comment (abannachGrafana) [inline actions/push-to-gcs/action.yaml:48]: are service accounts being automatically generated for each repo or is that up to the codeowners to create them in deployment tools?

Comment (iainlane) [reply actions/push-to-gcs/action.yaml:48]: the latter --- we have a TF module internally for that

**Comment by iainlane** [comment]
We need to align this with the internal implementation (the CSP resources we create for each repository). I'm going to this PR as a draft until that side is ready.

File: actions/push-to-gcs/action.yaml
Comment (dsotirakis) [inline actions/push-to-gcs/action.yaml:60]: you mean instead of the path?

File: actions/push-to-gcs/action.yaml
Comment (dsotirakis) [inline actions/push-to-gcs/action.yaml:23]: I like this `glob` thing - I wonder if it would make sense to make it part of the `path` so it's more clear to everybody?
Like check if there are `*` or `**` and have a step in this github action which will either resolve to a single object or a glob pattern? No strong opinions, I think it will just make things a little bit easier. Feel free to ignore if you have already thought about it and it works better your way!

File: actions/push-to-gcs/README.md
Comment (dsotirakis) [inline actions/push-to-gcs/README.md:75]: Hm I thought of that as well during the GAR work - I think for bucket names it makes more sense to just use names instead of ids. What I want to say by that is that for example we use `github-<repoID>-<env>@blah` for service accounts, but service accounts are something that won't bother us at all or will only bother us in some extreme/edge cases. In contrast, there will be times when we would like to manually check something from a bucket asap, and it would be more obvious to have something like `grafanalabs-grafana-enterprise-dev` rather than `grafanalabs-123456789-dev`. It's for practical reasons mostly. But still - not strongly opinionated. We can always revisit if discovery is hard!

**Review by dsotirakis — COMMENTED** [review]
Just a few things - I think it looks good! 💪

**Review by iainlane — COMMENTED** [review]
There's something that I've been thinking about for a while regarding our actions like these (build + push to GAR [we actually do have what I'm suggesting below for Docker Hub already], now GCS)...

I think we should provide the plain "login" functionality as a separate action, so we offer two things:

1. All in a box like we have here now: push these files for me
2. Get me set up to push and I'll handle the pushing

In other words, what do you think about splitting this into _two_ composite actions:

1. `authenticate-to-gcs-for-repo` which does all the steps apart from `google-github-actions/upload-cloud-storage` and the path manipulating stuff. It would output the bucket name to give to...
2. `push-to-gcs` which calls `authenticate-to-gcs-for-repo` and then `google-github-actions/upload-cloud-storage` after doing any necessary fiddling with paths

Here's the reason: I already know that we have people interacting with GCS from code. They don't need the action to do the pushing for them. We don't support those users when we only provide an opinionated action.

WDYT?

File: actions/push-to-gcs/action.yaml
Comment (ricky-undeadcoders) [reply actions/push-to-gcs/action.yaml:60]: Ah, good catch. I meant `object`, but I renamed `object` to `path` when I introduced the `glob` option. I will fix.

File: actions/push-to-gcs/README.md
Comment (ricky-undeadcoders) [reply actions/push-to-gcs/README.md:75]: Yeah, I know, I went this route because if you change the name of a GH repo you also have to recreate the bucket. This avoids that; but at this point both you and Iain have requested it... so it might be time to redo it.

I agree the user experience is worse with this route.

**Comment by ricky-undeadcoders** [comment]
> There's something that I've been thinking about for a while regarding our actions like these (build + push to GAR [we actually do have what I'm suggesting below for Docker Hub already], now GCS)...
> 
> I think we should provide the plain "login" functionality as a separate action, so we offer two things:
> 
>     1. All in a box like we have here now: push these files for me
> 
>     2. Get me set up to push and I'll handle the pushing
> 
> 
> In other words, what do you think about splitting this into _two_ composite actions:
> 
>     1. `authenticate-to-gcs-for-repo` which does all the steps apart from `google-github-actions/upload-cloud-storage` and the path manipulating stuff. It would output the bucket name to give to...
> 
>     2. `push-to-gcs` which calls `authenticate-to-gcs-for-repo` and then `google-github-actions/upload-cloud-storage` after doing any necessary fiddling with paths
> 
> 
> Here's the reason: I already know that we have people interacting with GCS from code. They don't need the action to do the pushing for them. We don't support those users when we only provide an opinionated action.
> 
> WDYT?

@iainlane Sounds good to me! I briefly reflected that it had to login 8 times to push my 8 different test cases. I'll get that done.

File: actions/push-to-gcs/README.md
Comment (ricky-undeadcoders) [reply actions/push-to-gcs/README.md]: This has been completed.

File: actions/push-to-gcs/action.yaml
Comment (ricky-undeadcoders) [reply actions/push-to-gcs/action.yaml:22]: I took a stab at implementing your recommendation, and I found it to be more confusing than leaving them separate. It did however give me the opportunity to revisit and simplify by eliminating the logic I was applying to those inputs, so now we simply pass the `path` and `glob` inputs directly to the downstream action. 

In the README I documented several cases for how to use these inputs in conjunction with each other. Let me know what you think!

File: actions/login-to-gcs/README.md
Comment (iainlane) [inline actions/login-to-gcs/README.md:29]: Suggest adding an example of using the bucket here - maybe can show that `gcloud` cli works?

```console
$ gcloud storage cp OBJECT_LOCATION gs://DESTINATION_BUCKET_NAME
```

as in: you don't have to use the `google-github-actions/upload-cloud-storage`

File: actions/login-to-gcs/action.yaml
Comment (iainlane) [inline actions/login-to-gcs/action.yaml:57]: nit: these steps could be combined

File: actions/login-to-gcs/action.yaml
Comment (iainlane) [inline actions/login-to-gcs/action.yaml:33]: also a nit, something that I've found useful:

```suggestion
        echo "project=${PROJECT}" | tee -a ${GITHUB_OUTPUT}
```

(also for the other outputs). It means the variables get logged which can be helpful sometimes.

File: actions/login-to-gcs/action.yaml
Comment (iainlane) [inline actions/login-to-gcs/action.yaml:27]: nit: noticed the two use different comparison operators

```suggestion
        elif [ "${{ inputs.environment }}" = "prod" ]; then
```

File: actions/login-to-gcs/action.yaml
Comment (iainlane) [inline actions/login-to-gcs/action.yaml:32]: suggestion: alternatively this could be written like so

```suggestion
      run: |
        case "${{ inputs.environment }}" in
          dev)
            PROJECT="grafanalabs-dev"
            ;;
          prod)
            PROJECT="grafanalabs-global"
            ;;
          *)
            echo "Invalid environment. Valid environment variable inputs: dev, prod"
            exit 1
            ;;
        esac
```

which might be a bit clearer to read - you decide :slightly_smiling_face:

File: actions/login-to-gcs/action.yaml
Comment (jdbaldry) [reply actions/login-to-gcs/action.yaml:27]: If we are using bash which https://github.com/grafana/shared-workflows/pull/40/files/ba125891a150ef68376d045749690d376437b07d#diff-8ad545e598c1ecec7544d085676f0f79de19ebdb2dc8309213d1ae2f8144497eR23 implies, we should use the builtin `[[ <TEST> ]]` form and `==`. Only if we are targeting POSIX sh should we use the test `[` binary and `=` equality comparison.

Comment (iainlane) [reply actions/login-to-gcs/action.yaml:27]: why _should_?

Comment (jdbaldry) [reply actions/login-to-gcs/action.yaml:27]: The bash builtin is faster, has more features, and more consistent syntax but ultimately in these cases they work the same. 

I think rules like this that encourage consistency also improve readability.

Comment (iainlane) [reply actions/login-to-gcs/action.yaml:27]: okey dokey. no particular objection - just that if we're going to ask people to abide by something as a rule then let's write it down if not enforce by CI.

we could defer to [the google guide](https://google.github.io/styleguide/shellguide.html) which does cover this case. If you want to make this a requirement, could you raise a PR please?

Comment (ricky-undeadcoders) [reply actions/login-to-gcs/action.yaml:27]: Yep, this was a silly bug. Fixed.

@jdbaldry I'm rather agnostic about how people choose to write their CI scripts (if it works it works), but if you'd like to put together recommendations in a PR we'll be happy to review. Maybe as part of the [GHA user docs](https://github.com/grafana/deployment_tools/blob/master/docs/platform/continuous-integration/github-actions-user-docs.md)?

File: actions/login-to-gcs/action.yaml
Comment (ricky-undeadcoders) [reply actions/login-to-gcs/action.yaml:32]: I'm happy with how it is because it matches what's in the related `push-to-gar-docker` action. What I would actually like to see here is another shared workflow that both the `gcs` and `gar` can call so we don't have to duplicate code, but I'm afraid that might be a poor design choice given that there's a limit to how many actions can be invoked.

In any case, if we were going to go that route I would like to attack it in another PR.

File: actions/login-to-gcs/action.yaml
Comment (jdbaldry) [reply actions/login-to-gcs/action.yaml:27]: Yeah, I don't have a problem with "if it works, it works". My drive-by comment was mostly to share knowledge about preferred practice rather than required.

That being said, given the esoteric syntaxes and number of quirks of shell scripting, I think you can get a lot from using tooling to lint and teach. I think shellcheck does a really good job of helping you learn how the shell is working so that you don't think something is working when it isn't.

File: actions/login-to-gcs/README.md
Comment (iainlane) [reply actions/login-to-gcs/README.md:26]: cheers - I would add that as a step in the example, but that's not required :slightly_smiling_face:

File: actions/login-to-gcs/action.yaml
Comment (iainlane) [reply actions/login-to-gcs/action.yaml:27]: @ricky-undeadcoders  If we're using `==` as recommended, use `[[` and `]]` too to use `bash`'s version of `test`.

File: actions/login-to-gcs/action.yaml
Comment (iainlane) [inline actions/login-to-gcs/action.yaml:37]:
```suggestion
```

File: actions/login-to-gcs/action.yaml
Comment (iainlane) [inline actions/login-to-gcs/action.yaml:47]:
```suggestion
        if [[ "${{ inputs.bucket }}" == "" ]]; then
```

File: actions/push-to-gcs/action.yaml
Comment (iainlane) [inline actions/push-to-gcs/action.yaml:54]:
```suggestion
        if [[ "${{ inputs.environment }}" == "dev" ]]; then
          PROJECT="grafanalabs-dev"
        elif [[ "${{ inputs.environment }}" == "prod" ]]; then
```

File: actions/push-to-gcs/action.yaml
Comment (iainlane) [inline actions/push-to-gcs/action.yaml:60]:
```suggestion
        echo "project=${PROJECT}" | tee -a ${GITHUB_OUTPUT}
```

File: actions/push-to-gcs/action.yaml
Comment (iainlane) [inline actions/push-to-gcs/action.yaml:78]:
```suggestion
        echo "destination=${DESTINATION}" | tee -a ${GITHUB_OUTPUT}
```

for consistency, and if you do this then the `echo` is not needed since `tee -a` is achieving the same thing :slightly_smiling_face:

File: actions/login-to-gcs/action.yaml
Comment (iainlane) [reply actions/login-to-gcs/action.yaml]: righto!

File: actions/push-to-gcs/README.md
Comment (iainlane) [inline actions/push-to-gcs/README.md:18]:
```suggestion
```

(not needed to add to `with` on the calls since this is the default)

File: actions/push-to-gcs/README.md
Comment (iainlane) [reply actions/push-to-gcs/README.md:23]: ok fair enough!

**Review by iainlane — APPROVED** [review]
Is that 404 in CI because this repository doesn't have a GAR? I suspect so - and this PR looks good to me, so feel free to fix that internally, then retry CI and once it passes merge. :+1: 

If that's not it you can dismiss this review

Thanks for all the work!

**Pull request merged**
